fix(oauth2):correct  header delimiter to use comma

dingjiayi opened the pull request:
<!--
NOTE: Please read the CONTRIBUTING.md guidelines before submitting your patch,
and ensure you followed them all:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributing

Refer to the Kong Gateway Community Pledge to understand how we work
with the open source community:
https://github.com/Kong/kong/blob/master/COMMUNITY_PLEDGE.md
-->

### Summary

Fix issue #14429 where authentication plugins were using spaces instead of commas to separate parameters in the `WWW-Authenticate` header. According to RFC 7235 (HTTP Authentication) ([Section 4.1](https://datatracker.ietf.org/doc/html/rfc7235#section-4.1)), the parameters should be comma-delimited.

### Related PRs:


### Checklist

- [ ] The Pull Request has tests
- [ ] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

<!--- If it fixes an open issue, please link to the issue here. -->
Fix #14429

samugi reviewed (commented):
looks good to me, it just need an entry in `changelog/unreleased`

dingjiayi commented:
> changelog/unreleased

@samugi Thank you for your review and feedback.  I've made the changes:

1. Added a helper function `make_www_authenticate_header` to improve code organization and make the header formatting more maintainable. 

2. Added the changelog entry.

chronolaw commented on kong/plugins/oauth2/access.lua:
```diff
@@ -89,6 +91,29 @@ do
   end
 end
 
+-- Helper function to construct WWW-Authenticate header
```
Could we use a `do end` block to minimize the scope like:
```
local make_www_authenticate_header
do
  local table_insert = table.insert
  local table_concat = table.concat
  local table_isempty = require("table.isempty")

  make_www_authenticate_header = function(...)
    ...
  end
end
```

chronolaw commented:
And if possible could we use `string.buffer` instead of `table.concat`?

chronolaw commented on changelog/unreleased/kong/fix-oauth2-header-delimiter.yml:
```diff
@@ -0,0 +1,3 @@
+message: "**OAuth2**: correct `WWW-Authenticate` header delimiter to use comma according to RFC 7235.."
```
```suggestion
message: "**OAuth2**: correct `WWW-Authenticate` header delimiter to use comma according to RFC 7235."
```

chronolaw commented on kong/plugins/oauth2/access.lua:
```diff
@@ -89,6 +91,36 @@ do
+      table_insert(params, fmt('error_description="%s"', error_description))
+    end
+
+    if #params > 0 then
```
should be `not table_isempty()`

dingjiayi replied:
changed.

dingjiayi commented:
> And if possible could we use `string.buffer` instead of `table.concat`?

Hi @chronolaw

Thank you for suggesting the use of `string.buffer` as an alternative to `table.concat` in the `make_www_authenticate_header` function. I've carefully evaluated this suggestion by implementing and benchmarking both approaches.

After thorough testing, I don't recommend making this change for the following reasons:

1. There's no clear benefit in terms of code readability or maintainability with `string.buffer` for this particular use case. 

2. Performance testing revealed a significant difference between the two approaches. The `string.buffer` implementation runs at approximately 1/3 the speed of the current `table.concat` implementation.

Here's a summary of the benchmark results from my local testing (100M iterations):
<img width="696" alt="image" src="https://github.com/user-attachments/assets/b197629f-fd57-4aa5-ad31-2be6e54142be" />


```lua
-- table.concat implementation (current)
local ngx = ngx
local fmt = string.format
local table_insert = table.insert
local table_concat = table.concat

local function make_www_authenticate_header(realm, error_code, error_description)
  local params = {}

  if realm and realm ~= "" then
    table_insert(params, fmt('realm="%s"', realm))
  end

  if error_code then
    table_insert(params, fmt('error="%s"', error_code))
  end

  if error_description then
    table_insert(params, fmt('error_description="%s"', error_description))
  end

  if #params > 0 then
    return "Bearer " .. table_concat(params, ", ")
  else
    return "Bearer"
  end
end

ngx.update_time()
local start = ngx.now()
for i = 1, 100000000 do
  make_www_authenticate_header("test_realm", "test_error", "test_description")
end
ngx.update_time()
print("Execution time(by table_concat): " .. (ngx.now() - start) .. " seconds")
```

```lua
local buffer = require("string.buffer")

local function make_www_authenticate_header(realm, error_code, error_description)
  
  local buf = buffer.new()
  buf:put("Bearer")
  if realm and realm ~= "" then
    buf:putf(" realm=%s", realm)
  end

  if error_code then
    buf:putf(", error=%s", error_code)
  end

  if error_description then
    buf:putf(", error_description=%s", error_description)
  end

  return buf:get()
end


ngx.update_time()
local start = ngx.now()
for i = 1, 100000000 do
  make_www_authenticate_header("test_realm", "test_error", "test_description")
end
ngx.update_time()
print("Execution time(by string.buffer): " .. (ngx.now() - start) .. " seconds")
```

If there are any considerations I may have overlooked or if my testing methodology could be improved, please let me know. I'm happy to conduct additional tests or explore other implementation approaches as needed.

Thank you for your review and suggestions!

chronolaw commented:
That's OK, thank you. And I am not very familiar with this plugin, we need other people to review the logic.

nowNick reviewed (approved):
This looks really good! Thank you @dingjiayi for taking additional effort to refactor this code a little bit 🏅

samugi reviewed (approved):
looks good to me
